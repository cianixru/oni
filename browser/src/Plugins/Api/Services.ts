import { getInstance, VersionControlManager } from "./../../Services/VersionControl"

export class Services {
    public get vcs(): VersionControlManager {
        return getInstance()
    }
}
