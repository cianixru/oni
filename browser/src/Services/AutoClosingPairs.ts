/**
 * AutoClosingPairs
 *
 * Service to enable auto-closing pair key bindings
 */

import * as Oni from "oni-api"

import { Configuration } from "./Configuration"
import { EditorManager } from "./EditorManager"
import { InputManager } from "./InputManager"
import { LanguageManager } from "./Language"

import { NeovimInstance } from "./../neovim"

import * as Log from "./../Log"

export interface IAutoClosingPair {
    open: string
    close: string
    // TODO: Support `notIn` equivalent
}

export const activate = (
    configuration: Configuration,
    editorManager: EditorManager,
    inputManager: InputManager,
    languageManager: LanguageManager,
) => {
    const insertModeFilter = () => editorManager.activeEditor.mode === "insert"

    let subscriptions: Oni.DisposeFunction[] = []

    const handleOpenCharacter = (pair: IAutoClosingPair, editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            // TODO: PERFORMANCE: Look at how to collapse this instead of needed multiple asynchronous calls.
            await inputFunc(pair.open + pair.close)

            const pos = await neovim.callFunction("getpos", ["."])
            const [, oneBasedLine, oneBasedColumn] = pos
            await editor.activeBuffer.setCursorPosition(oneBasedLine - 1, oneBasedColumn - 2)
        })

        return true
    }

    const handleBackspaceCharacter = (pairs: IAutoClosingPair[], editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer
            const lines = await activeBuffer.getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
            )
            const line = lines[0]

            const { column } = activeBuffer.cursor

            const matchingPair = pairs.find(p => {
                return column >= 1 && line[column] === p.close && line[column - 1] === p.open
            })

            if (matchingPair) {
                // Remove the pairs
                const beforePair = line.substring(0, column - 1)
                const afterPair = line.substring(column + 1, line.length)

                const pos = await neovim.callFunction("getpos", ["."])
                const [, oneBasedLine, oneBasedColumn] = pos
                await editor.activeBuffer.setCursorPosition(oneBasedLine - 1, oneBasedColumn - 2)

                await activeBuffer.setLines(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.line + 1,
                    [beforePair + afterPair],
                )
            } else {
                await inputFunc("<bs>")
            }
        })

        return true
    }

    const handleEnterCharacter = (pairs: IAutoClosingPair[], editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer

            const lines = await (activeBuffer as any).getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
                false,
            )
            const line = lines[0]

            const { column } = activeBuffer.cursor

            const matchingPair = pairs.find(p => {
                return column >= 1 && line[column] === p.close && line[column - 1] === p.open
            })

            if (matchingPair) {
                const whiteSpacePrefix = getWhiteSpacePrefix(line)
                const beforePair = line.substring(0, column)
                const afterPair = line.substring(column, line.length)

                const pos = await neovim.callFunction("getpos", ["."])
                const [, oneBasedLine] = pos
                await activeBuffer.setLines(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.line + 1,
                    [beforePair, whiteSpacePrefix, whiteSpacePrefix + afterPair],
                )
                await activeBuffer.setCursorPosition(oneBasedLine, whiteSpacePrefix.length)
                await inputFunc("<tab>")
            } else {
                await inputFunc("<enter>")
            }
        })

        return true
    }

    const handleCloseCharacter = (pair: IAutoClosingPair, editor: Oni.Editor) => () => {
        const neovim: any = editor.neovim
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer
            const lines = await (activeBuffer as any).getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
                false,
            )
            const line = lines[0]
            if (line[activeBuffer.cursor.column] === pair.close) {
                await activeBuffer.setCursorPosition(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.column + 1,
                )
            } else {
                await inputFunc(pair.close)
            }
        })

        return true
    }

    const onBufferEnter = (newBuffer: Oni.Buffer) => {
        if (!configuration.getValue("autoClosingPairs.enabled")) {
            Log.verbose("[Auto Closing Pairs] Not enabled.")
            return
        }

        if (subscriptions && subscriptions.length) {
            subscriptions.forEach(df => df())
        }

        subscriptions = []

        const autoClosingPairs = getAutoClosingPairs(configuration, newBuffer.language)

        autoClosingPairs.forEach(pair => {
            subscriptions.push(
                inputManager.bind(
                    pair.open,
                    handleOpenCharacter(pair, editorManager.activeEditor),
                    insertModeFilter,
                ),
            )
            subscriptions.push(
                inputManager.bind(
                    pair.close,
                    handleCloseCharacter(pair, editorManager.activeEditor),
                    insertModeFilter,
                ),
            )
        })

        subscriptions.push(
            inputManager.bind(
                "<bs>",
                handleBackspaceCharacter(autoClosingPairs, editorManager.activeEditor),
                insertModeFilter,
            ),
        )
        subscriptions.push(
            inputManager.bind(
                "<enter>",
                handleEnterCharacter(autoClosingPairs, editorManager.activeEditor),
                insertModeFilter,
            ),
        )
    }

    editorManager.activeEditor.onBufferEnter.subscribe(onBufferEnter)

    const activeEditor = editorManager.activeEditor
    if (activeEditor && activeEditor.activeBuffer) {
        onBufferEnter(activeEditor.activeBuffer)
    }
}

const nonWhiteSpaceRegEx = /\S/

export const getWhiteSpacePrefix = (str: string): string => {
    const firstIndex = str.search(nonWhiteSpaceRegEx)

    if (firstIndex === -1) {
        return ""
    } else {
        return str.substring(0, firstIndex)
    }
}

const getAutoClosingPairs = (
    configuration: Configuration,
    language: string,
): IAutoClosingPair[] => {
    const languagePairs = configuration.getValue(`language.${language}.autoClosingPairs`)

    if (languagePairs) {
        return languagePairs
    } else {
        return configuration.getValue("autoClosingPairs.default") || []
    }
}
