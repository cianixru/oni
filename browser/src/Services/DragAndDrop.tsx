import * as React from "react"
import * as DND from "react-dnd"

type Render<T> = (props: T) => React.ReactElement<T>
type OnDrop = (item: any) => object | void
type IsValidDrop = (item: any) => boolean

// Drop Target ================================================================
export interface IDroppeable {
    isOver?: boolean
    connectDropTarget?: any
    onDrop: OnDrop
    isValidDrop: IsValidDrop
    canDrop?: boolean
    didDrop?: boolean
    accepts: string[] | string
    render: Render<{
        canDrop: boolean
        isOver?: boolean
        didDrop?: boolean
        connectDropTarget?: DND.DropTargetConnector
    }>
}
interface DroppedProps {
    onDrop: OnDrop
    isValidDrop: IsValidDrop
}

const DropTarget = {
    drop(dropped: DroppedProps, monitor: DND.DropTargetMonitor) {
        return dropped.onDrop({ drag: monitor.getItem(), drop: dropped })
    },
    canDrop(props: DroppedProps, monitor: DND.DropTargetMonitor) {
        return props.isValidDrop({ drag: monitor.getItem(), drop: props })
    },
}

const DropCollect = (connect: DND.DropTargetConnector, monitor: DND.DropTargetMonitor) => ({
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop(),
    didDrop: monitor.didDrop(),
})

/**
 * A component which can have items of a specific matching type dropped onto it
 */
@DND.DropTarget<IDroppeable>(({ accepts }) => accepts, DropTarget, DropCollect)
export class Droppeable<P extends IDroppeable> extends React.Component<P> {
    public render() {
        const { isOver, connectDropTarget, canDrop, didDrop } = this.props
        return connectDropTarget(<div>{this.props.render({ isOver, canDrop, didDrop })}</div>)
    }
}

// Drag Source ================================================================
export interface IDraggeable {
    target?: string
    isDragging?: boolean
    connectDragSource?: any
    render: Render<{ isDragging?: boolean; connectDragSource?: DND.DragSourceConnector }>
}

const DragSource = {
    beginDrag(props: object) {
        return props
    },
}

const DragCollect = (connect: DND.DragSourceConnector, monitor: DND.DragSourceMonitor) => {
    return {
        connectDragSource: connect.dragSource(),
        isDragging: monitor.isDragging(),
    }
}

/**
 * A component that can be dragged onto a droppeable one
 *
 * @name props
 * @function
 * @param {String | String[]} >props.target The target Type that responds to the drop
 * @param {Object} DragSource Object with a beginDrag which return the dragged props
 * @param {React.Component} A component which is dragged onto another
 * @returns {React.Component<P>} A react class component
 */
@DND.DragSource<IDraggeable>(props => props.target, DragSource, DragCollect)
export class Draggeable<P extends IDraggeable> extends React.Component<P> {
    public render() {
        const { isDragging, connectDragSource } = this.props
        return connectDragSource(<div>{this.props.render({ isDragging })}</div>)
    }
}

interface IDragDrop {
    isOver?: boolean
    didDrop?: boolean
    onDrop: OnDrop
    accepts: string[] | string
    connectDropTarget?: any
    canDrop?: boolean
    isValidDrop: IsValidDrop
    dragTarget: string
    isDragging?: boolean
    connectDragSource?: any
    render: Render<{ didDrop?: boolean; canDrop?: boolean; isOver?: boolean; isDragging?: boolean }>
}

/**
 * A render prop which takes a given component and makes it a drop target as well as draggeable
 */
@DND.DropTarget<IDragDrop>(props => props.accepts, DropTarget, DropCollect)
@DND.DragSource<IDragDrop>(props => props.dragTarget, DragSource, DragCollect)
export class DragAndDrop<P extends IDragDrop> extends React.Component<P> {
    public render() {
        const { connectDragSource, connectDropTarget } = this.props

        return connectDropTarget(connectDragSource(<div>{this.props.render(this.props)}</div>))
    }
}
