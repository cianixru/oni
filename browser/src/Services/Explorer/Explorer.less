@import (reference) "./../../UI/components/common.less";

.explorer {
    transform: translateX(-3px);
    transition: transform 0.25s ease;
    user-select: none;
    cursor: default;

    display: flex;
    flex-direction: column;

    .header {
        flex: 0 0 auto;
    }

    .loaded & {
        transform: translateX(0px);
    }

    .items {
        flex: 1 1 auto;
        overflow-y: auto;

        .item {
            display: flex;
            flex-direction: row;
            justify-content: center;
            align-items: center;
            padding: 4px;

            .icon {
                flex: 0 0 auto;
                width: 20px;
                text-align: center;
            }

            .name {
                flex: 1 1 auto;
            }
        }
    }
}
