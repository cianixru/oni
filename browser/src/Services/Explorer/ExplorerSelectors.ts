/**
 * ExplorerSelectors.ts
 *
 * Selectors for the explorer state
 */

import * as path from "path"

import * as flatten from "lodash/flatten"

import { ExpandedFolders, FolderOrFile, IExplorerState } from "./ExplorerStore"

export interface IContainerNode {
    id: string
    type: "container"
    expanded: boolean
    name: string
}

export interface IFolderNode {
    id: string
    type: "folder"
    folderPath: string
    expanded: boolean
    name: string
    indentationLevel: number
}

export interface IFileNode {
    id: string
    type: "file"
    filePath: string
    modified: boolean
    name: string
    indentationLevel: number
}

export const EmptyNode: ExplorerNode = {
    type: null,
    id: null,
    modified: null,
    filePath: null,
    name: null,
    indentationLevel: null,
}

export type ExplorerNode = IContainerNode | IFolderNode | IFileNode

export const isPathExpanded = (state: IExplorerState, pathToCheck: string): boolean => {
    return !!state.expandedFolders[pathToCheck]
}

export const mapStateToNodeList = (state: IExplorerState): ExplorerNode[] => {
    let ret: ExplorerNode[] = []

    //     ret.push({
    //         id: "opened",
    //         type: "container",
    //         expanded: true,
    //         name: "Opened Files",
    //     })

    //     const openedFiles: ExplorerNode[] = Object.keys(state.openedFiles)
    //         .filter(filePath => !!filePath)
    //         .map(filePath => ({
    //         type: "file",
    //         id: "opened:" + filePath,
    //         filePath,
    //         name: path.basename(filePath),
    //         modified: false, // TODO
    //         indentationLevel: 0,
    //     } as ExplorerNode))

    //     ret = [...ret, ...openedFiles]

    if (!state.rootFolder || !state.rootFolder.fullPath) {
        return ret
    }

    ret.push({
        id: "explorer",
        type: "container",
        expanded: !!state.expandedFolders[state.rootFolder.fullPath],
        name: state.rootFolder.fullPath,
    })

    const expandedTree = flattenFolderTree(state.rootFolder, [], state.expandedFolders, 0)

    // The root node is included in the output, so we'll remove it
    const [, ...remainingTree] = expandedTree

    ret = [...ret, ...remainingTree]
    return ret
}

export const flattenFolderTree = (
    folderTree: FolderOrFile,
    currentList: ExplorerNode[],
    expandedFolders: ExpandedFolders,
    indentationLevel: number,
): ExplorerNode[] => {
    switch (folderTree.type) {
        case "file":
            const file: ExplorerNode = {
                type: "file",
                name: path.basename(folderTree.fullPath),
                id: "explorer:" + folderTree.fullPath,
                filePath: folderTree.fullPath,
                modified: false,
                indentationLevel,
            }
            return [...currentList, file]
        case "folder":
            const expanded = !!expandedFolders[folderTree.fullPath]

            const folder: ExplorerNode = {
                type: "folder",
                id: "explorer:" + folderTree.fullPath,
                folderPath: folderTree.fullPath,
                name: path.basename(folderTree.fullPath),
                expanded,
                indentationLevel,
            }

            const folderChildren = expandedFolders[folderTree.fullPath] || []
            const children = flatten(
                folderChildren.map(c =>
                    flattenFolderTree(c, [], expandedFolders, indentationLevel + 1),
                ),
            )

            return [...currentList, folder, ...children]
        default:
            return []
    }
}
