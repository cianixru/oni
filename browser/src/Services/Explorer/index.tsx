/**
 * Explorer/index.tsx
 *
 * Entry point for explorer-related features
 */

import { CommandManager } from "./../CommandManager"
import { EditorManager } from "./../EditorManager"
import { SidebarManager } from "./../Sidebar"
import { Workspace } from "./../Workspace"

import { ExplorerSplit } from "./ExplorerSplit"

export const activate = (
    commandManager: CommandManager,
    editorManager: EditorManager,
    sidebarManager: SidebarManager,
    workspace: Workspace,
) => {
    sidebarManager.add("files-o", new ExplorerSplit(workspace, commandManager, editorManager))
}
