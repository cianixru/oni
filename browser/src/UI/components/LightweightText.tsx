import * as React from "react"
import { connect } from "react-redux"

import * as State from "./../Shell/ShellState"

import { focusManager } from "./../../Services/FocusManager"

export interface ITextInputViewProps {
    onComplete?: (result: string) => void
    onChange?: (evt: React.ChangeEvent<HTMLInputElement>) => void

    defaultValue?: string

    backgroundColor: string
    foregroundColor: string

    overrideDefaultStyle?: boolean
}

// TODO: Is there a better value for this?
const WordRegex = /[$_a-zA-Z0-9]/i
const EmptyStyle: React.CSSProperties = {}

/**
 * TextInputView is a lightweight input control, that implements some
 * common functionality (like focus management, key handling)
 */
export class TextInputView extends React.PureComponent<ITextInputViewProps, {}> {
    private _element: HTMLInputElement

    public componentDidMount(): void {
        if (this._element) {
            focusManager.pushFocus(this._element)
        }
    }

    public render(): JSX.Element {
        const containerStyle: React.CSSProperties = this.props.overrideDefaultStyle
            ? EmptyStyle
            : {
                  padding: "4px",
                  border: "1px solid " + this.props.foregroundColor,
              }

        const inputStyle: React.CSSProperties = {
            outline: "none",
            color: this.props.foregroundColor,
            backgroundColor: this.props.backgroundColor,
            border: "0px",
            transform: "translateY(0px)",
        }

        const defaultValue = this.props.defaultValue || ""

        return (
            <div style={containerStyle}>
                <input
                    type="text"
                    style={inputStyle}
                    placeholder={defaultValue}
                    onKeyDown={evt => this._onKeyDown(evt)}
                    onChange={evt => this._onChange(evt)}
                    onFocus={evt => evt.currentTarget.select()}
                    ref={elem => (this._element = elem)}
                />
            </div>
        )
    }

    public componentWillUnmount(): void {
        if (this._element) {
            if (this.props.onComplete) {
                this.props.onComplete(this._element.value)
            }

            focusManager.popFocus(this._element)
            this._element = null
        }
    }

    private _onChange(changeEvent: React.ChangeEvent<HTMLInputElement>): void {
        if (this.props.onChange) {
            this.props.onChange(changeEvent)
        }
    }

    private _onKeyDown(keyboardEvent: React.KeyboardEvent<HTMLInputElement>): void {
        if (this._element && keyboardEvent.ctrlKey) {
            switch (keyboardEvent.key) {
                case "u": {
                    this._element.value = ""
                    break
                }
                case "h": {
                    const previousValue = this._element.value

                    if (previousValue.length > 0) {
                        this._element.value = previousValue.substring(0, previousValue.length - 1)
                    }
                    break
                }
                case "w":
                    {
                        const previousValue = this._element.value
                        let idx = previousValue.length - 1

                        while (idx > 0) {
                            if (!previousValue[idx].match(WordRegex)) {
                                break
                            }

                            idx--
                        }

                        this._element.value = previousValue.substring(0, idx)
                    }
                    break
                default:
                    return
            }
        }
    }
}

const mapStateToProps = (state: State.IState, originalProps?: Partial<ITextInputViewProps>) => ({
    ...originalProps,
    backgroundColor: state.colors["editor.background"],
    foregroundColor: state.colors["editor.foreground"],
})

export const TextInput = connect(mapStateToProps)(TextInputView)
