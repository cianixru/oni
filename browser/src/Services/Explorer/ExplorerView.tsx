/**
 * ExplorerSplit.tsx
 *
 */

import * as React from "react"
import { connect } from "react-redux"

import styled from "styled-components"

import { SidebarContainerView, SidebarItemView } from "./../../UI/components/SidebarItemView"
import { VimNavigator } from "./../../UI/components/VimNavigator"

import { FileIcon } from "./../FileIcon"

import * as ExplorerSelectors from "./ExplorerSelectors"
import { IExplorerState } from "./ExplorerStore"

export interface INodeViewProps {
    node: ExplorerSelectors.ExplorerNode
    isSelected: boolean
    onClick: () => void
}

const NodeWrapper = styled.div`
    &:hover {
        text-decoration: underline;
    }
`

// tslint:disable-next-line
const noop = (elem: HTMLElement) => {}
const scrollIntoViewIfNeeded = (elem: HTMLElement) => {
    // tslint:disable-next-line
    elem && elem["scrollIntoViewIfNeeded"] && elem["scrollIntoViewIfNeeded"]()
}

export class NodeView extends React.PureComponent<INodeViewProps, {}> {
    public render(): JSX.Element {
        return (
            <NodeWrapper
                style={{ cursor: "pointer" }}
                onClick={() => this.props.onClick()}
                innerRef={this.props.isSelected ? scrollIntoViewIfNeeded : noop}
            >
                {this.getElement()}
            </NodeWrapper>
        )
    }

    public getElement(): JSX.Element {
        const node = this.props.node

        switch (node.type) {
            case "file":
                return (
                    <SidebarItemView
                        text={node.name}
                        isFocused={this.props.isSelected}
                        isContainer={false}
                        indentationLevel={node.indentationLevel}
                        icon={<FileIcon fileName={node.name} isLarge={true} />}
                    />
                )
            case "container":
                return (
                    <SidebarContainerView
                        isContainer={true}
                        isExpanded={node.expanded}
                        text={node.name}
                        isFocused={this.props.isSelected}
                    />
                )
            case "folder":
                return (
                    <SidebarContainerView
                        isContainer={false}
                        isExpanded={node.expanded}
                        text={node.name}
                        isFocused={this.props.isSelected}
                        indentationLevel={node.indentationLevel}
                    />
                )
            default:
                return <div>{JSON.stringify(node)}</div>
        }
    }
}

export interface IExplorerViewContainerProps {
    onSelectionChanged: (id: string) => void
    onClick: (id: string) => void
}

export interface IExplorerViewProps extends IExplorerViewContainerProps {
    nodes: ExplorerSelectors.ExplorerNode[]
    isActive: boolean
}

import { SidebarEmptyPaneView } from "./../../UI/components/SidebarEmptyPaneView"
import { Sneakable } from "./../../UI/components/Sneakable"

import { commandManager } from "./../CommandManager"

export class ExplorerView extends React.PureComponent<IExplorerViewProps, {}> {
    public render(): JSX.Element {
        const ids = this.props.nodes.map(node => node.id)

        if (!this.props.nodes || !this.props.nodes.length) {
            return (
                <SidebarEmptyPaneView
                    active={this.props.isActive}
                    contentsText="Nothing to show here, yet!"
                    actionButtonText="Open a Folder"
                    onClickButton={() => commandManager.executeCommand("workspace.openFolder")}
                />
            )
        }

        return (
            <VimNavigator
                ids={ids}
                active={this.props.isActive}
                onSelectionChanged={this.props.onSelectionChanged}
                onSelected={id => this.props.onClick(id)}
                render={(selectedId: string) => {
                    const nodes = this.props.nodes.map(node => (
                        <Sneakable callback={() => this.props.onClick(node.id)} key={node.id}>
                            <NodeView
                                node={node}
                                isSelected={node.id === selectedId}
                                onClick={() => this.props.onClick(node.id)}
                            />
                        </Sneakable>
                    ))

                    return (
                        <div className="explorer enable-mouse">
                            <div className="items">{nodes}</div>
                        </div>
                    )
                }}
            />
        )
    }
}

const mapStateToProps = (
    state: IExplorerState,
    containerProps: IExplorerViewContainerProps,
): IExplorerViewProps => {
    return {
        ...containerProps,
        isActive: state.hasFocus,
        nodes: ExplorerSelectors.mapStateToNodeList(state),
    }
}

export const Explorer = connect(mapStateToProps)(ExplorerView)
