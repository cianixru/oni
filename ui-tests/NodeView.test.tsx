import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import * as React from "react"

import { DragAndDrop } from "./../browser/src/Services/DragAndDrop"
import { ExplorerNode } from "./../browser/src/Services/Explorer/ExplorerSelectors"
import { NodeView, NodeWrapper } from "./../browser/src/Services/Explorer/ExplorerView"
import { TextInputView } from "./../browser/src/UI/components/LightweightText"

describe("<NodeView />", () => {
    const testNode = {
        id: "2",
        filePath: "/test/a/file.txt",
        type: "file",
        modified: false,
        name: "file.txt",
        indentationLevel: 2,
    } as ExplorerNode

    const Node = (
        <NodeView
            yanked={[] as string[]}
            measure={() => null}
            isCreating={false}
            isRenaming={testNode}
            moveFileOrFolder={() => ({})}
            node={testNode}
            isSelected={false}
            onClick={() => ({})}
            onCancelRename={() => ({})}
            onCompleteRename={() => ({})}
        />
    )
    it("Should render without crashing", () => {
        const wrapper = shallow(Node)
        expect(wrapper.length).toEqual(1)
    })

    it("Should render and input element if the selected element is renaming", () => {
        const wrapper = shallow(
            <NodeView
                yanked={[] as string[]}
                measure={() => null}
                isCreating={false}
                isRenaming={testNode}
                moveFileOrFolder={() => ({})}
                node={testNode}
                isSelected={true}
                onClick={() => ({})}
                onCancelRename={() => ({})}
                onCompleteRename={() => ({})}
            />,
        )

        const input = wrapper.find(TextInputView)
        expect(input.length).toEqual(1)
    })

    it("Should render the node if the selected element is not renaming", () => {
        const wrapper = shallow(Node)
        expect(
            wrapper
                .find(NodeWrapper)
                .dive()
                .find(DragAndDrop),
        ).toHaveLength(1)
    })

    it("Should match the snapshot", () => {
        const wrapper = shallow(Node)
        expect(shallowToJson(wrapper)).toMatchSnapshot()
    })
})
