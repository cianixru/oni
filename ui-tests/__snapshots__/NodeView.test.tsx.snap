// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeView /> Should match the snapshot 1`] = `
<styled.div
  innerRef={[Function]}
  style={
    Object {
      "cursor": "pointer",
    }
  }
>
  <div>
    <DropTarget(DragSource(DragAndDrop))
      accepts={
        Array [
          "FILE",
          "FOLDER",
        ]
      }
      dragTarget="FILE"
      isValidDrop={[Function]}
      node={
        Object {
          "filePath": "/test/a/file.txt",
          "id": "2",
          "indentationLevel": 2,
          "modified": false,
          "name": "file.txt",
          "type": "file",
        }
      }
      onDrop={[Function]}
      render={[Function]}
    />
  </div>
</styled.div>
`;
